common/flatpak-dir.c

```diff
@@ -6953,10 +6953,36 @@ flatpak_dir_remove_ref (FlatpakDir   *self,
                         GCancellable *cancellable,
                         GError      **error)
 {
-  if (!ostree_repo_set_ref_immediate (self->repo, remote_name, ref, NULL, cancellable, error))
+  if (flatpak_dir_use_system_helper (self, NULL))
+    {
+      const char *installation = flatpak_dir_get_id (self);
+      FlatpakSystemHelper *system_helper = flatpak_dir_get_system_helper (self);
+
+      /* If we don't have the system helper, we'll have to try and just remove
+       * the ref as an unprivileged user, which might fail later */
+      if (system_helper)
+        {
+          if (!flatpak_system_helper_call_remove_local_ref_sync (system_helper,
+                                                                 remote_name,
+                                                                 ref,
+                                                                 installation ? installation : "",
+                                                                 cancellable,
+                                                                 error))
+            return FALSE;
+        }
+
+      return TRUE;
+    }
+
+  if (!ostree_repo_set_ref_immediate (self->repo,
+                                      remote_name,
+                                      ref,
+                                      NULL,
+                                      cancellable,
+                                      error))
     return FALSE;
 
-  return TRUE;
+  return flatpak_dir_prune (self, cancellable, error);
```

Comment:
This will always prune, which is already done in e.g. flatpak_dir_uninstall() which calls remove_ref(). This means it will be done twice, and additionally, with the new prune locks, i think this will deadlock.

Also, i think it will not be entirely uncommon to remove several refs and in that case it will be more useful to prune only at the end.

Reply:
Yeah. I think what I will do here is pull the `flatpak_dir_prune` to the library boundary (so if you remove a single local ref, the prune happens immediately afterwards, otherwise the prune happens once all refs are removed).

Or should we make it the caller's responsibility and expose it in the library?